src/Compilers/CSharp/Portable/Syntax/CSharpNonRecursiveSyntaxRewriter.cs

```diff
@@ -0,0 +1,253 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// Represents a non-recursive visitor which descends an entire <see cref="CSharpSyntaxNode"/> graph and
+    /// may replace or remove visited SyntaxNodes in depth-first order.
+    /// </summary>
+    public class CSharpNonRecursiveSyntaxRewriter : CSharpSyntaxVisitor<SyntaxNode>
+    {
+        private readonly Stack<SyntaxNodeOrToken> _undeconstructedStack;
+        private readonly Stack<UntransformedNode> _untransformedStack;
+        private readonly Stack<SyntaxNodeOrToken> _transformedStack;
+        private readonly NodeDeconstructor _deconstructor;
+        private readonly NodeReassembler _reassembler;
+        private readonly TriviaRewriter _trivializer;
+        private readonly List<SyntaxNodeOrToken> _children;
+
+        protected CSharpNonRecursiveSyntaxRewriter(bool visitIntoStructuredTrivia = false)
+        {
+            _undeconstructedStack = new Stack<SyntaxNodeOrToken>();
+            _transformedStack = new Stack<SyntaxNodeOrToken>();
+            _untransformedStack = new Stack<UntransformedNode>();
+            _deconstructor = new NodeDeconstructor();
+            _reassembler = new NodeReassembler();
+            _trivializer = new TriviaRewriter(this, visitIntoStructuredTrivia);
+            _children = new List<SyntaxNodeOrToken>();
+        }
+
+        protected SyntaxNode Original { get; private set; }
+
+        protected virtual bool Skip(SyntaxNodeOrToken nodeOrToken, out SyntaxNodeOrToken rewriten)
```

Comment:
I'm still curious about the out parameter.  I'm not sure when this gets used.  Also, there is no ShouldRewriteToken?

---
In reply to: [73028078](https://github.com/dotnet/roslyn/pull/12494#discussion_r73028078) [](ancestors = 73028078)